### PR TITLE
Image Resolver for purchasable entities

### PR DESCRIPTION
**Pull request by mmoreram**
* Created a set of service for resolution of a main image
* Covers product, variant and pack. All in one service called elcodi.image_resolver.purchasable
* Created Twig extension for template
* Added some extra fixtures data for testing + Some functional testing
* Tested 100% coverage on Unit Testing
* Added a new interface that combines both ImageContainer and PrincipalImageContainer features (for mocking purposes)

### CODE REVIEW

File: src/Elcodi/Bundle/MediaBundle/Resources/config/services.yml [hunk 1]
```diff
@@ -23,3 +23,6 @@ services:
             - '@elcodi.manager.media_file'
             - '@elcodi.manager.media_image'
             - '@elcodi.event_dispatcher.media'
+
+    elcodi.resolver.image:
+        class: Elcodi\Component\Media\Services\ImageResolver
```
Comment (EmanueleMinotto) [inline src/Elcodi/Bundle/MediaBundle/Resources/config/services.yml]: add a newline

File: src/Elcodi/Bundle/ProductBundle/DataFixtures/ORM/Traits/ProductWithImagesTrait.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+     * @param ImagesContainerWithPrincipalImageInterface $imageContainer Image Container
+     * @param string                                     $imageName      Image name
+     *
+     * @return $this Self object
+     */
+    protected function storeProductImage(
+        ImagesContainerWithPrincipalImageInterface $imageContainer,
+        $imageName
+    ) {
+        $imagePath = realpath(dirname(__FILE__) . '/../images/' . $imageName);
```
Comment (EmanueleMinotto) [inline src/Elcodi/Bundle/ProductBundle/DataFixtures/ORM/Traits/ProductWithImagesTrait.php]: you can replace `dirname(__FILE__)` with `__DIR__`

File: src/Elcodi/Component/Product/ImageResolver/ProductImageResolver.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        return 'Elcodi\Component\Product\Entity\Interfaces\ProductInterface';
+    }
+
+    /**
+     * Get valid Image.
+     *
+     * @param PurchasableInterface $purchasable Purchasable
+     *
+     * @return ImageInterface Image resolved
```
Comment (EmanueleMinotto) [inline src/Elcodi/Component/Product/ImageResolver/ProductImageResolver.php]: `@return ImageInterface|false`

File: src/Elcodi/Component/Product/ImageResolver/VariantImageResolver.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        return 'Elcodi\Component\Product\Entity\Interfaces\VariantInterface';
+    }
+
+    /**
+     * Get valid Image.
+     *
+     * @param PurchasableInterface $purchasable Purchasable
+     *
+     * @return ImageInterface Image resolved
```
Comment (EmanueleMinotto) [inline src/Elcodi/Component/Product/ImageResolver/VariantImageResolver.php]: same as above

File: src/Elcodi/Component/Product/ImageResolver/Abstracts/AbstractImageResolver.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        $this->purchasableImageResolvers[] = $purchasableImageResolver;
+    }
+
+    /**
+     * Get valid Image.
+     *
+     * @param PurchasableInterface $purchasable Purchasable
+     *
+     * @return ImageInterface Image resolved
```
Comment (EmanueleMinotto) [inline src/Elcodi/Component/Product/ImageResolver/Abstracts/AbstractImageResolver.php]: This returns an ImageInterface or a boolean

**Comment by mmoreram** [comment]
@EmanueleMinotto updated :) Thanks!

**Comment by EmanueleMinotto** [comment]
:+1:

**Pull request merged**